runtime/src/mailbox_api.rs

```diff
@@ -0,0 +1,147 @@
+// Licensed under the Apache-2.0 license
+
+use caliptra_drivers::{CaliptraError, CaliptraResult};
+use zerocopy::{AsBytes, FromBytes, LayoutVerified};
+
+#[derive(PartialEq, Eq)]
+pub struct CommandId(pub u32);
+
+impl CommandId {
+    pub const FIRMWARE_LOAD: Self = Self(0x46574C44); // "FWLD"
+    pub const GET_IDEV_CSR: Self = Self(0x49444556); // "IDEV"
+    pub const GET_LDEV_CERT: Self = Self(0x4C444556); // "LDEV"
+    pub const ECDSA384_VERIFY: Self = Self(0x53494756); // "SIGV"
+    pub const STASH_MEASUREMENT: Self = Self(0x4D454153); // "MEAS"
+    pub const INVOKE_DPE: Self = Self(0x44504543); // "DPEC"
+
+    pub const TEST_ONLY_GET_LDEV_CERT: Self = Self(0x4345524c); // "CERL"
+    pub const TEST_ONLY_GET_FMC_ALIAS_CERT: Self = Self(0x43455246); // "CERF"
+
+    /// FIPS module commands.
+    /// The status command.
+    pub const VERSION: Self = Self(0x4650_5652); // "FPVR"
+    /// The self-test command.
+    pub const SELF_TEST: Self = Self(0x4650_4C54); // "FPST"
+    /// The shutdown command.
+    pub const SHUTDOWN: Self = Self(0x4650_5344); // "FPSD"
+}
+impl From<u32> for CommandId {
+    fn from(value: u32) -> Self {
+        Self(value)
+    }
+}
+impl From<CommandId> for u32 {
+    fn from(value: CommandId) -> Self {
+        value.0
+    }
+}
+
+// Helpers
+// Assumes given byte array is >= size_of T
+// Will return CaliptraError::RUNTIME_INTERNAL otherwise
+pub fn cast_bytes_to_struct<T: AsBytes + FromBytes>(bytes: &[u8]) -> CaliptraResult<&T> {
+    Ok(
+        LayoutVerified::<&[u8], T>::new(&bytes[..core::mem::size_of::<T>()])
+        .ok_or(CaliptraError::RUNTIME_INTERNAL)?
+        .into_ref()
+    )
+}
+pub fn cast_bytes_to_struct_mut<T: AsBytes + FromBytes>(bytes: &mut [u8]) -> CaliptraResult<&mut T> {
+    Ok(
+        LayoutVerified::<&mut [u8], T>::new(&mut bytes[..core::mem::size_of::<T>()])
+        .ok_or(CaliptraError::RUNTIME_INTERNAL)?
+        .into_mut()
+    )
+}
+
+// COMMON
+#[repr(C)]
+#[derive(AsBytes, FromBytes)]
+pub struct MailboxReqCommon {
+    pub chksum: i32,
+}
+
+#[repr(C)]
+#[derive(AsBytes, FromBytes)]
+pub struct MailboxRespCommon {
+    pub chksum: i32,
+    pub fips_status: u32,
+}
+
+pub const FIPS_STATUS_APPROVED: u32 = 0;
+
+// CALIPTRA_FW_LOAD
+#[repr(C)]
+#[derive(AsBytes, FromBytes)]
+pub struct CaliptraFwLoadReq {
+    pub common: MailboxReqCommon,
+    pub data: [u8; 0], // variable length
```

Comment:
I would recommend just choosing a static value (probably 1024). Especially since these are just test commands.

Declaring variable arrays like this doesn't work in rust, since it would be unsafe.

Reply:
Thanks, this was something I was meaning to ask about. Should 1024 be safe for all the messages with variable length fields? (CALIPTRA_FW_LOAD request, GET_IDEV_CSR response, GET_LDEV_CERT response, INVOKE_DPE_COMMAND request and response)

Reply:
`CALIPTRA_FW_LOAD` shouldn't be copied from the mailbox at all. We don't have enough stack to hold the full firmware image. On reset, ROM will read the data from the mailbox into ICCM.

For GET_IDEV_CSR response and GET_LDEV_CERT response, 1K should be fine.

INVOKE_DPE_COMMAND, we're still working it out. I think the response buffer may need to be 4KiB. But we can figure it out later when that command is added.

Reply:
That makes more sense, I was a little unclear on that for a while.

Sounds good for the rest.